Retry starting/installing service upon failure

ycombinator opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR. Please explain the WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
-->

This PR enhances the service runtime to retry starting a service if it fails to start, up to a maximum number of attempts.  Starting a service also installs it first.

## Why is it important?

<!-- Mandatory
Explain here the WHY, or the rationale/motivation for the changes.
-->

To make service startup resilient to transient errors.

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [ ] I have read and understood the [pull request guidelines](https://github.com/elastic/elastic-agent/blob/main/CONTRIBUTING.md#pull-request-guidelines) of this project.
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [ ] I have added an integration test or an E2E test

## Disruptive User Impact

<!--
Will the changes introduced by this PR cause disruption to users in any way? If so, please describe what changes users
could make on their end to nullify or minimize this disruption. Consider impacts in related systems, not just directly
when using Elastic Agent.
-->

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
-

## Questions to ask yourself

- How are we going to support this in production?
- How are we going to measure its adoption?
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

<!-- CI Cheatsheet
Trigger comments:
/test             (Or `buildkite test this|it`) Triggers unit test pipeline
/test extended    (Or `buildkite test extended`) Triggers integration test pipeline

PR labels:
skip-ci           Skips unit and integration tests
skip-it           Skips integration tests
-->

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/25076)
* Commit: 4455e4a322485a5ee86c83c138b77e041fe2447f

cc @ycombinator

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25076","number":25076,"commit":"4455e4a322485a5ee86c83c138b77e041fe2447f"}],"number":25076}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/25334)
* Commit: 5e38b725ff7e5f99a7291ed024b243f1eebf03fd

### Failed CI Steps
* [Unit tests - Windows 2022](https://buildkite.com/elastic/elastic-agent/builds/25334#0198af13-da65-4f52-952b-a5a280f88b65)
* [Unit tests - Windows 2022](https://buildkite.com/elastic/elastic-agent/builds/25334#0198af60-93b2-485c-a5a7-43e04c700643)
* [Unit tests - Windows 2016](https://buildkite.com/elastic/elastic-agent/builds/25334#0198af13-da65-42c6-bf5d-fef8d7fd8dab)
* [Unit tests - Windows 10](https://buildkite.com/elastic/elastic-agent/builds/25334#0198af13-da69-43c9-9e64-06788f8c9925)
* [Unit tests - Windows 11](https://buildkite.com/elastic/elastic-agent/builds/25334#0198af13-da69-49a7-85ae-aafdec40717d)

### History
* :green_heart: [Build #25076](https://buildkite.com/elastic/elastic-agent/builds/25076) succeeded 4455e4a322485a5ee86c83c138b77e041fe2447f

cc @ycombinator

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25334","number":25334,"commit":"5e38b725ff7e5f99a7291ed024b243f1eebf03fd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25076","number":25076,"commit":"4455e4a322485a5ee86c83c138b77e041fe2447f"}],"number":25334}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/25470)
* Commit: c3e7ca6392e8467adef6afb5ea693013d78df0d9

### Failed CI Steps
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/25470#0198c256-9bb4-4d47-9e26-93b77f495136)
* [Unit tests - macOS 15 ARM](https://buildkite.com/elastic/elastic-agent/builds/25470#0198c270-07cc-423f-94d9-33fc06ab8042)

### History
* :broken_heart: [Build #25334](https://buildkite.com/elastic/elastic-agent/builds/25334) failed 5e38b725ff7e5f99a7291ed024b243f1eebf03fd
* :green_heart: [Build #25076](https://buildkite.com/elastic/elastic-agent/builds/25076) succeeded 4455e4a322485a5ee86c83c138b77e041fe2447f

cc @ycombinator

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25470","number":25470,"commit":"c3e7ca6392e8467adef6afb5ea693013d78df0d9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25334","number":25334,"commit":"5e38b725ff7e5f99a7291ed024b243f1eebf03fd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25076","number":25076,"commit":"4455e4a322485a5ee86c83c138b77e041fe2447f"}],"number":25470}
buildkite-pr-comment-->

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/25475)
* Commit: f7479a8f7463ade23e1991335d341243cdf00cf6

### Failed CI Steps
* [Unit tests - Windows 2022](https://buildkite.com/elastic/elastic-agent/builds/25475#0198c2c2-58a1-4dda-a629-e6f755c4521b)
* [Unit tests - Windows 2022](https://buildkite.com/elastic/elastic-agent/builds/25475#0198c311-f12a-485e-870b-1a328631a575)
* [Unit tests - Windows 2016](https://buildkite.com/elastic/elastic-agent/builds/25475#0198c2c2-58a2-459f-b69e-8474d5688a74)
* [Unit tests - Windows 10](https://buildkite.com/elastic/elastic-agent/builds/25475#0198c2c2-58ae-4e96-aef4-0ceb4e272d6a)
* [Unit tests - Windows 11](https://buildkite.com/elastic/elastic-agent/builds/25475#0198c2c2-58ae-4cfb-983c-a65d71c1067f)

### History
* :broken_heart: [Build #25470](https://buildkite.com/elastic/elastic-agent/builds/25470) failed c3e7ca6392e8467adef6afb5ea693013d78df0d9
* :broken_heart: [Build #25334](https://buildkite.com/elastic/elastic-agent/builds/25334) failed 5e38b725ff7e5f99a7291ed024b243f1eebf03fd
* :green_heart: [Build #25076](https://buildkite.com/elastic/elastic-agent/builds/25076) succeeded 4455e4a322485a5ee86c83c138b77e041fe2447f

cc @ycombinator

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25475","number":25475,"commit":"f7479a8f7463ade23e1991335d341243cdf00cf6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25470","number":25470,"commit":"c3e7ca6392e8467adef6afb5ea693013d78df0d9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25334","number":25334,"commit":"5e38b725ff7e5f99a7291ed024b243f1eebf03fd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25076","number":25076,"commit":"4455e4a322485a5ee86c83c138b77e041fe2447f"}],"number":25475}
buildkite-pr-comment-->

blakerouse reviewed (commented):
This is an improvement to the service manager, but the issue I have with this is even with retries onces the `maxServiceStartAttempts` value is reached it still won't try to start the service. Being that the Elastic Agent is a service manager when it comes to a service component, it really should keep trying.

What do I mean by keep trying? I mean that after some missed check-ins that the Elastic Agent should try to start it again. This is exactly how the `command` runtime works, and I think a similar approach should be taken here.

In the command runtime the actionState is stored here: https://github.com/elastic/elastic-agent/blob/main/pkg/component/runtime/command.go#L160.

Then during the `<-t.C` it ensures that it is running here: https://github.com/elastic/elastic-agent/blob/main/pkg/component/runtime/command.go#L231

I think a similar approach is needed here. The code should try to start the service, but if it fails to start then it should retry (as you have it), if it was never successful to start then during the `<-t.C` is should try to start it again. Then if it is successful, but is missing 3 check-ins its probably best to do another start to ensure that it is running still.

ycombinator commented:
Note to self: test if this PR would also mitigate https://github.com/elastic/elastic-agent/issues/7412.

elasticmachine commented:
Pinging @elastic/elastic-agent-control-plane (Team:Elastic-Agent-Control-Plane)

ycombinator commented:
Thanks, @blakerouse, for pointing out the parallels with the command runtime implementation.  Very helpful and it makes sense for the two runtimes to be consistent in their (re)start behavior.

> the issue I have with this is even with retries onces the `maxServiceStartAttempts` value is reached it still won't try to start the service. Being that the Elastic Agent is a service manager when it comes to a service component, it really should keep trying.

Removed `maxServiceStartAttempts` in a8bd97c4be73b16f80a579be6f07c02665b95336.

> the code should try to start the service, but if it fails to start then it should retry (as you have it), if it was never successful to start then during the `<-t.C` is should try to start it again. Then if it is successful, but is missing 3 check-ins its probably best to do another start to ensure that it is running still.

Implemented this in b1dd80b1b42dafca146bd4fb41c7f12ffe11fbe3.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/25881)
* Commit: b1dd80b1b42dafca146bd4fb41c7f12ffe11fbe3

### Failed CI Steps
* [Unit tests - Windows 2022](https://buildkite.com/elastic/elastic-agent/builds/25881#0198f260-f7be-4363-bc34-d8538fb01656)
* [Unit tests - Windows 2022](https://buildkite.com/elastic/elastic-agent/builds/25881#0198f2b4-3d93-4ed6-b733-5e7ac859c877)
* [Unit tests - Windows 2016](https://buildkite.com/elastic/elastic-agent/builds/25881#0198f260-f7bf-45c3-aa61-3a67ee846e76)
* [Unit tests - Windows 10](https://buildkite.com/elastic/elastic-agent/builds/25881#0198f260-f7c3-43ed-85fc-d9dd45014d6b)
* [Unit tests - Windows 11](https://buildkite.com/elastic/elastic-agent/builds/25881#0198f260-f7c4-4bb6-bf20-c00e9a332e3a)
* [Win2025:sudo:install-uninstall](https://buildkite.com/elastic/elastic-agent/builds/25881#0198f296-9b29-421e-aa71-ab20a0e5f5cf)

### History
* :green_heart: [Build #25563](https://buildkite.com/elastic/elastic-agent/builds/25563) succeeded f6cd4dc7ddbef0409ef35d7c8141e8cc0b3607f8
* :broken_heart: [Build #25475](https://buildkite.com/elastic/elastic-agent/builds/25475) failed f7479a8f7463ade23e1991335d341243cdf00cf6
* :broken_heart: [Build #25470](https://buildkite.com/elastic/elastic-agent/builds/25470) failed c3e7ca6392e8467adef6afb5ea693013d78df0d9
* :broken_heart: [Build #25334](https://buildkite.com/elastic/elastic-agent/builds/25334) failed 5e38b725ff7e5f99a7291ed024b243f1eebf03fd
* :green_heart: [Build #25076](https://buildkite.com/elastic/elastic-agent/builds/25076) succeeded 4455e4a322485a5ee86c83c138b77e041fe2447f

cc @ycombinator

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25881","number":25881,"commit":"b1dd80b1b42dafca146bd4fb41c7f12ffe11fbe3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25563","number":25563,"commit":"f6cd4dc7ddbef0409ef35d7c8141e8cc0b3607f8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25475","number":25475,"commit":"f7479a8f7463ade23e1991335d341243cdf00cf6"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25470","number":25470,"commit":"c3e7ca6392e8467adef6afb5ea693013d78df0d9"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25334","number":25334,"commit":"5e38b725ff7e5f99a7291ed024b243f1eebf03fd"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25076","number":25076,"commit":"4455e4a322485a5ee86c83c138b77e041fe2447f"}],"number":25881}
buildkite-pr-comment-->

blakerouse commented on pkg/component/runtime/service.go:
```diff
@@ -38,6 +39,9 @@ var (
 	// ErrInvalidServiceSpec error invalid service specification.
 	ErrInvalidServiceSpec = errors.New("invalid service spec")
+
+	serviceRestartDelay   = 30 * time.Second // variable so it can be shortened in tests
```
instead of setting this as a variable in the whole module could this be moved to be a variable in the struct and then set this value in the struct in the tests?

blakerouse replied:
Looks good.

blakerouse reviewed (dismissed):
Looks good. Thanks for refactoring this!

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/elastic-agent/builds/26123)
* Commit: 8cd931badb80ff5646c2afa65386c801478dc0b7

### History
* :broken_heart: [Build #25930](https://buildkite.com/elastic/elastic-agent/builds/25930) failed 8cd931badb80ff5646c2afa65386c801478dc0b7
* :broken_heart: [Build #25921](https://buildkite.com/elastic/elastic-agent/builds/25921) failed d1aadaead8d545e2ac8e9ca5b598ca274db16df7
* :broken_heart: [Build #25881](https://buildkite.com/elastic/elastic-agent/builds/25881) failed b1dd80b1b42dafca146bd4fb41c7f12ffe11fbe3
* :green_heart: [Build #25563](https://buildkite.com/elastic/elastic-agent/builds/25563) succeeded f6cd4dc7ddbef0409ef35d7c8141e8cc0b3607f8
* :broken_heart: [Build #25475](https://buildkite.com/elastic/elastic-agent/builds/25475) failed f7479a8f7463ade23e1991335d341243cdf00cf6

cc @ycombinator

<!--buildkite-pr-comment-elastic-agent
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/26123","number":26123,"commit":"8cd931badb80ff5646c2afa65386c801478dc0b7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25930","number":25930,"commit":"8cd931badb80ff5646c2afa65386c801478dc0b7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25921","number":25921,"commit":"d1aadaead8d545e2ac8e9ca5b598ca274db16df7"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25881","number":25881,"commit":"b1dd80b1b42dafca146bd4fb41c7f12ffe11fbe3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/elastic-agent/builds/25563","number":25563,"commit":"f6cd4dc7ddbef0409ef35d7c8141e8cc0b3607f8"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/elastic-agent/builds/25475","number":25475,"commit":"f7479a8f7463ade23e1991335d341243cdf00cf6"}],"number":26123}
buildkite-pr-comment-->
